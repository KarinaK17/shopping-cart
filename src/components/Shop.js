import React, { useEffect, useState } from "react";
import bluePens from "../images/pens3.jpg";
import goldenPens from "../images/pens1.jpg";
import colorfulPens from "../images/pens2.jpg";
import pencils from "../images/pencil.jpg";
import calculator from "../images/calculator.jpg";
import stickyNotes from "../images/sticky-notes.jpg";
import highlighters from "../images/highlighters.jpg";
import landscapeCopybooks from "../images/landscape-copybooks.jpg";
import natureCopybooks from "../images/nature-copybooks.jpg";
import multicoloredPencils from "../images/multicolored-pencils.jpg";
import brushes from "../images/brushes.jpg";
import watercolor from "../images/watercolor.jpg";
import Cart from "./Cart";
import ProductCard from "./ProductCard";

const Shop = () => {
  const [products, setProducts] = useState([
    {
      name: "blue pens",
      price: 4,
      quantity: 0,
      src: bluePens,
      description: "Blue ink ballpoint pens, pack of 2",
    },
    {
      name: "golden pens",
      price: 6,
      quantity: 0,
      src: goldenPens,
      description: "Automatic oil-based blue ink ballpoint pens, pack of 2",
    },
    {
      name: "colorful pens",
      price: 8,
      quantity: 0,
      src: colorfulPens,
      description: "Blue oil-based ink ballpoint pens (multicolor), pack of 5",
    },
    {
      name: "pencils",
      price: 4,
      quantity: 0,
      src: pencils,
      description: 'Graphite pencils (hardness "B"), pack of 3',
    },
    {
      name: "claculator",
      price: 15,
      quantity: 0,
      src: calculator,
      description: "Claculator (size: 195*143*37 mm)",
    },
    {
      name: "sticky notes",
      price: 3,
      quantity: 0,
      src: stickyNotes,
      description: "Block of multicolor sticky notes, 100 sheets",
    },
    {
      name: "highlighters",
      price: 10,
      quantity: 0,
      src: highlighters,
      description:
        "Fluorescent water-based ink multicolor highlighters, pack of 4",
    },
    {
      name: "landscape copybooks",
      price: 10,
      quantity: 0,
      src: landscapeCopybooks,
      description: "A5 format notebooks (48 sheet), pack of 5",
    },
    {
      name: "nature copybooks",
      price: 10,
      quantity: 0,
      src: natureCopybooks,
      description: "A5 format notebooks (48 sheet), pack of 5",
    },
    {
      name: "multicolored pencils",
      price: 10,
      quantity: 0,
      src: multicoloredPencils,
      description:
        "Pencils with multicolor soft lead, 7 colors per pencil, pack of 3",
    },
    {
      name: "brushes",
      price: 12,
      quantity: 0,
      src: brushes,
      description: "Artistic synthetic brushes, pack of 3",
    },
    {
      name: "watercolor",
      price: 24,
      quantity: 0,
      src: watercolor,
      description: "High-quality honey-based watercolor paint, 24 colors",
    },
  ]);

  const [total, setTotal] = useState({ number: 0, sum: 0 });

  const handleQuantityChange = (event, name) => {
    let p = products.map((x) => {
      if (x.name === name) {
        if (
          isNaN(event.target.value) ||
          event.target.value === "" ||
          Number(event.target.value) < 0
        ) {
          console.log(isNaN(x.quantity));
          return { ...x, quantity: 0 };
        } else {
          return { ...x, quantity: Number(event.target.value) };
        }
      } else {
        return x;
      }
    });
    setProducts(p);
  };

  const handleCartClick = (name) => {
    let p = products.map((x) => {
      if (x.name === name) {
        return { ...x, quantity: 1 };
      } else {
        return x;
      }
    });
    setProducts(p);
  };

  const handleIncrement = (name) => {
    let p = products.map((x) => {
      if (x.name === name) {
        return { ...x, quantity: x.quantity + 1 };
      } else {
        return x;
      }
    });
    setProducts(p);
  };

  const handleDecrement = (name) => {
    let p = products.map((x) => {
      if (x.name === name) {
        return { ...x, quantity: x.quantity - 1 };
      } else {
        return x;
      }
    });
    setProducts(p);
  };

  useEffect(() => {
    const numberOfProducts = products.reduce((accumulator, object) => {
      return accumulator + object.quantity;
    }, 0);
    const sumOfProducts = products.reduce((accumulator, object) => {
      return accumulator + object.quantity * object.price;
    }, 0);
    setTotal({ number: numberOfProducts, sum: sumOfProducts });

    // let p = products.map((x) => {
    //   if (isNaN(x.quantity) || x.quantity === "") {
    //     console.log(isNaN(x.quantity));
    //     return { ...x, quantity: 0 };
    //   } else {
    //     return x;
    //   }
    // });
    // setProducts(p);
  }, [products]);

  return (
    <div className="content Shop">
      <Cart number={total.number} sum={total.sum} />
      <h2>Shop</h2>
      <div className="products-container">
        {products.map((product) => {
          return (
            <ProductCard
              key={product.name}
              index={products.indexOf(product)}
              name={product.name}
              src={product.src}
              description={product.description}
              quantity={product.quantity}
              price={product.price}
              handleQuantityChange={handleQuantityChange}
              handleCartClick={handleCartClick}
              handleIncrement={handleIncrement}
              handleDecrement={handleDecrement}
            />
          );
        })}
      </div>
    </div>
  );
};

export default Shop;
