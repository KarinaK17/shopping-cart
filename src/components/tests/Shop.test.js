import { render, screen, act } from "@testing-library/react";
import React from "react";
import userEvent from "@testing-library/user-event";
import Shop from "../Shop";

describe("Shop component", () => {
  it("renders correctly", () => {
    const { container } = render(<Shop />);
    expect(container).toMatchSnapshot();
  });

  it("changes cart to input if you click on cart", async () => {
    const user = userEvent.setup();
    render(<Shop />);

    const cart = screen.getByTestId("cart-0");
    await act(async () => {
      await user.click(cart);
    });

    expect(screen.queryByTestId("cart-0")).not.toBeInTheDocument();
    expect(screen.getByTestId("input-0")).toBeInTheDocument();
    expect(screen.getByTestId("input-0")).toHaveValue("1");
  });

  it("plus button works", async () => {
    const user = userEvent.setup();
    render(<Shop />);

    const cart = screen.getByTestId("cart-0");
    await act(async () => {
      await user.click(cart);
    });

    const plus = screen.getByText("+");

    await act(async () => {
      await user.click(plus);
    });

    expect(screen.queryByTestId("cart-0")).not.toBeInTheDocument();
    expect(screen.getByTestId("input-0")).toBeInTheDocument();
    expect(screen.getByTestId("input-0")).toHaveValue("2");
  });

  it("minus button works (and input changes to cart if quantity is less than 1)", async () => {
    const user = userEvent.setup();
    render(<Shop />);

    const cart = screen.getByTestId("cart-0");
    await act(async () => {
      await user.click(cart);
    });

    const minus = screen.getByText("â€“");

    await act(async () => {
      await user.click(minus);
    });

    expect(screen.getByTestId("cart-0")).toBeInTheDocument();
    expect(screen.queryByTestId("input-0")).not.toBeInTheDocument();
  });

  it("total sum and number of items change if you add something to the cart", async () => {
    const user = userEvent.setup();
    render(<Shop />);

    const cart = screen.getByTestId("cart-11");
    await act(async () => {
      await user.click(cart);
    });

    const plus = screen.getByText("+");

    await act(async () => {
      await user.click(plus);
    });

    expect(screen.getByTestId("input-11")).toHaveValue("2");
    expect(screen.getByTestId("number-of-items")).toHaveTextContent("2 items");
    expect(screen.getByTestId("sum-of-items")).toHaveTextContent("Total: 48 $");
  });
});
