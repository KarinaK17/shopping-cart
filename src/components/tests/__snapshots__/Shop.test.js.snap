// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shop component renders correctly 1`] = `
<div>
  <div
    class="content Shop"
  >
    <div
      class="Cart"
    >
      <img
        alt="shopping cart"
        class="cart"
        src="white-cart.svg"
      />
      <p
        data-testid="number-of-items"
      >
        0
         items 
      </p>
      <p
        data-testid="sum-of-items"
      >
        Total: 
        0
         $
      </p>
    </div>
    <h2>
      Shop
    </h2>
    <div
      class="products-container"
    >
      <div
        class="Product-card"
      >
        <div
          class="image-container"
        >
          <img
            alt="blue pens"
            class="product-img"
            src="pens3.jpg"
          />
          <p
            class="product-description"
          >
            Blue ink ballpoint pens, pack of 2
          </p>
        </div>
        <p>
          Price: 
          4
           $
        </p>
        <div>
          <img
            alt="shopping cart"
            class="cart"
            data-testid="cart-0"
            src="cart.svg"
          />
        </div>
      </div>
      <div
        class="Product-card"
      >
        <div
          class="image-container"
        >
          <img
            alt="golden pens"
            class="product-img"
            src="pens1.jpg"
          />
          <p
            class="product-description"
          >
            Automatic oil-based blue ink ballpoint pens, pack of 2
          </p>
        </div>
        <p>
          Price: 
          6
           $
        </p>
        <div>
          <img
            alt="shopping cart"
            class="cart"
            data-testid="cart-1"
            src="cart.svg"
          />
        </div>
      </div>
      <div
        class="Product-card"
      >
        <div
          class="image-container"
        >
          <img
            alt="colorful pens"
            class="product-img"
            src="pens2.jpg"
          />
          <p
            class="product-description"
          >
            Blue oil-based ink ballpoint pens (multicolor), pack of 5
          </p>
        </div>
        <p>
          Price: 
          8
           $
        </p>
        <div>
          <img
            alt="shopping cart"
            class="cart"
            data-testid="cart-2"
            src="cart.svg"
          />
        </div>
      </div>
      <div
        class="Product-card"
      >
        <div
          class="image-container"
        >
          <img
            alt="pencils"
            class="product-img"
            src="pencil.jpg"
          />
          <p
            class="product-description"
          >
            Graphite pencils (hardness "B"), pack of 3
          </p>
        </div>
        <p>
          Price: 
          4
           $
        </p>
        <div>
          <img
            alt="shopping cart"
            class="cart"
            data-testid="cart-3"
            src="cart.svg"
          />
        </div>
      </div>
      <div
        class="Product-card"
      >
        <div
          class="image-container"
        >
          <img
            alt="claculator"
            class="product-img"
            src="calculator.jpg"
          />
          <p
            class="product-description"
          >
            Claculator (size: 195*143*37 mm)
          </p>
        </div>
        <p>
          Price: 
          15
           $
        </p>
        <div>
          <img
            alt="shopping cart"
            class="cart"
            data-testid="cart-4"
            src="cart.svg"
          />
        </div>
      </div>
      <div
        class="Product-card"
      >
        <div
          class="image-container"
        >
          <img
            alt="sticky notes"
            class="product-img"
            src="sticky-notes.jpg"
          />
          <p
            class="product-description"
          >
            Block of multicolor sticky notes, 100 sheets
          </p>
        </div>
        <p>
          Price: 
          3
           $
        </p>
        <div>
          <img
            alt="shopping cart"
            class="cart"
            data-testid="cart-5"
            src="cart.svg"
          />
        </div>
      </div>
      <div
        class="Product-card"
      >
        <div
          class="image-container"
        >
          <img
            alt="highlighters"
            class="product-img"
            src="highlighters.jpg"
          />
          <p
            class="product-description"
          >
            Fluorescent water-based ink multicolor highlighters, pack of 4
          </p>
        </div>
        <p>
          Price: 
          10
           $
        </p>
        <div>
          <img
            alt="shopping cart"
            class="cart"
            data-testid="cart-6"
            src="cart.svg"
          />
        </div>
      </div>
      <div
        class="Product-card"
      >
        <div
          class="image-container"
        >
          <img
            alt="landscape copybooks"
            class="product-img"
            src="landscape-copybooks.jpg"
          />
          <p
            class="product-description"
          >
            A5 format notebooks (48 sheet), pack of 5
          </p>
        </div>
        <p>
          Price: 
          10
           $
        </p>
        <div>
          <img
            alt="shopping cart"
            class="cart"
            data-testid="cart-7"
            src="cart.svg"
          />
        </div>
      </div>
      <div
        class="Product-card"
      >
        <div
          class="image-container"
        >
          <img
            alt="nature copybooks"
            class="product-img"
            src="nature-copybooks.jpg"
          />
          <p
            class="product-description"
          >
            A5 format notebooks (48 sheet), pack of 5
          </p>
        </div>
        <p>
          Price: 
          10
           $
        </p>
        <div>
          <img
            alt="shopping cart"
            class="cart"
            data-testid="cart-8"
            src="cart.svg"
          />
        </div>
      </div>
      <div
        class="Product-card"
      >
        <div
          class="image-container"
        >
          <img
            alt="multicolored pencils"
            class="product-img"
            src="multicolored-pencils.jpg"
          />
          <p
            class="product-description"
          >
            Pencils with multicolor soft lead, 7 colors per pencil, pack of 3
          </p>
        </div>
        <p>
          Price: 
          10
           $
        </p>
        <div>
          <img
            alt="shopping cart"
            class="cart"
            data-testid="cart-9"
            src="cart.svg"
          />
        </div>
      </div>
      <div
        class="Product-card"
      >
        <div
          class="image-container"
        >
          <img
            alt="brushes"
            class="product-img"
            src="brushes.jpg"
          />
          <p
            class="product-description"
          >
            Artistic synthetic brushes, pack of 3
          </p>
        </div>
        <p>
          Price: 
          12
           $
        </p>
        <div>
          <img
            alt="shopping cart"
            class="cart"
            data-testid="cart-10"
            src="cart.svg"
          />
        </div>
      </div>
      <div
        class="Product-card"
      >
        <div
          class="image-container"
        >
          <img
            alt="watercolor"
            class="product-img"
            src="watercolor.jpg"
          />
          <p
            class="product-description"
          >
            High-quality honey-based watercolor paint, 24 colors
          </p>
        </div>
        <p>
          Price: 
          24
           $
        </p>
        <div>
          <img
            alt="shopping cart"
            class="cart"
            data-testid="cart-11"
            src="cart.svg"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
